requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -0,0 +1,787 @@
+```
+
+There are many ways to keep this performant. For example an optimization pass on the AST that merges nodes into faster composed nodes. `funnels.steps[] -> events` becomes `funnels.steps_from_events[]`.
+
+Yet such an optimisation step is overkill at this stage. The easiest solution is to just peek inside the inputs when parsing them.
+
+That basically means, instead of
+
+```ts
+function parseFunnelDropoff(node: FunnelDropoffNode, ctx: QueryContext) {
+    const sql = ctx.wrap(opts.from)
+    return `query something (${sql}) step ${ctx.wrap(step)} power to the people`
+}
+function parseFunnel(node: FunnelDropoffNode, ctx: QueryContext) {
+    const sql = ctx.wrap(opts.steps.map((step) => `(${ctx.wrap(opts.step)})`))
+    return `make funnel from (${sql}) get stuff`
+}
+```
+
+you write
+
+```ts
+function parseFunnelDropoff(node: FunnelDropoffNode, ctx: QueryContext) {
+    const sql = ctx.wrap(opts.from.steps.map((step) => `(${ctx.wrap(opts.step)})`))
+    return `query something (${sql}) step ${ctx.wrap(step)} do other stuff`
+}
+```
+
+### Tree parsing
+
+I don't know what's the right shape for the parser. I don't know if this should be a bunch of functions, or classes. I do however know we need to pass around a global `ctx` object
+
+```ts
+// either this
+class FunnelDropoffNode extends Node {
+    toSQL({ wrap, ctx, opts }) {
+        // opts.from is FunnelNode
+        return `query something (${wrap(opts.from, ctx)}) step ${wrap(step, ctx)} do other stuff`
+    }
+}
+
+// or this
+function parseFunnelDropoff(node: FunnelDropoffNode, ctx: QueryContext) {
+    const sql = ctx.wrap(opts.from)
+    return `query something (${sql}) step ${ctx.wrap(step)} power to the people`
+}
+```
+
+This `ctx` could be used to carry global filters to the end, e.g. from `funnel` to the `events` themselves.
+
+```ts
+function parseFunnel(node: FunnelDropoffNode, ctx: QueryContext) {
+    ctx.filters = [...ctx.filters, ...node.globalFilters]
+    const sql = ctx.wrap(opts.steps.map((step) => `(${ctx.wrap(opts.step)})`))
+    return `make funnel from (${sql}) get stuff`
+}
+```
+
+This section requires a deeper dive, possibly during implementation.
+
+### Multiple levels
+
+As long as somewhere in the chain we have something that normalises to either `persons` or `events`, we're golden. These should be the common targets, that can act as input to a lot of queries and visualizations. Each level in this chain will require a subquery.
+
+### Pagination
+
+TODO
+
+### Versioning
+
+TODO. We will possibly need an `apiVersion` field in each AST node.
+
+## Next steps for the backend
+
+This is a gradual rollout. There will be no big rewrite. The steps below will also bring no visible changes on the frontend.
+
+1. Refactor the huge `FilterType` into `InsightTrendsFilter`, `InsightFunnelsFilter`, etc. It's still the same backend, just with cleaner queries (e.g. remove all `funnel*` fields for the trends filter).
```

Comment:
Non-blocking, but I think it's worth having some strategy for the migration between these different data models, especially as we save them as filters in dashboarditem which'll make the caching messy.